File: 10_greedy_algorithms/scala/01_set_covering.scala
import scala.collection.mutable

// You pass a collection in, and it gets converted to a set.
val statesNeeded = mutable.Set("mt", "wa", "or", "id", "nv", "ut", "ca", "az")

val stations = mutable.Map[String, mutable.Set[String]](
  "kone" -> mutable.Set("id", "nv", "ut"),
  "ktwo" -> mutable.Set("wa", "id", "mt"),
  "kthree" -> mutable.Set("or", "nv", "ca"),
  "kfour" -> mutable.Set("nv", "ut"),
  "kfive" -> mutable.Set("ca", "az")
)

def mySetCovering(statesNeeded: mutable.Set[String], stations: mutable.Map[String, mutable.Set[String]]): Option[mutable.Set[String]] = {
  val finalStations = mutable.Set[String]()
  val remainingStations = mutable.Map[String, mutable.Set[String]]() ++ stations
  
  while(statesNeeded.nonEmpty) {
    var bestStation: Option[String] = None
    var statesCovered = mutable.Set[String]()
    
    for((station, statesForStation) <- remainingStations) {
      val covered = statesNeeded.intersect(statesForStation)
      if(covered.size > statesCovered.size && !finalStations.contains(station)) {
        bestStation = Some(station)
        statesCovered = covered
      }
    }
    
    if(bestStation.isDefined) {
      statesNeeded --= statesCovered
      finalStations += bestStation.get
      remainingStations.remove(bestStation.get)
    } else {
      return None
    }
  }
  
  Some(finalStations)
}

println(mySetCovering(statesNeeded, stations))

File: 10_greedy_algorithms/clojure/01_set_covering.clj
(ns set-covering)

;; You pass a collection in, and it gets converted to a set.
(def states-needed #{"mt" "wa" "or" "id" "nv" "ut" "ca" "az"})

(def stations {"kone" #{"id" "nv" "ut"}
               "ktwo" #{"wa" "id" "mt"}
               "kthree" #{"or" "nv" "ca"}
               "kfour" #{"nv" "ut"}
               "kfive" #{"ca" "az"}})

(defn my-set-covering [states-needed stations]
  (loop [remaining-states states-needed
         remaining-stations stations
         final-stations #{}]
    (if (empty? remaining-states)
      final-stations
      (let [best-station (reduce
                          (fn [best [station states-for-station]]
                            (let [covered (clojure.set/intersection remaining-states states-for-station)
                                  best-covered (if best
                                                 (clojure.set/intersection remaining-states (get remaining-stations best))
                                                 #{})]
                              (if (and (not (contains? final-stations station))
                                       (> (count covered) (count best-covered)))
                                station
                                best)))
                          nil
                          remaining-stations)]
        (if best-station
          (let [states-covered (clojure.set/intersection remaining-states
                                                          (get remaining-stations best-station))]
            (recur (clojure.set/difference remaining-states states-covered)
                   (dissoc remaining-stations best-station)
                   (conj final-stations best-station)))
          nil)))))

(println (my-set-covering states-needed stations))

